### PR TITLE
FFmpeg integration (again) & peer testing preparations

### LINKED ISSUE #63: Deliverable #2 - Program generates a file, which maps the phonemes from the audio and text to timestamps
**Is this related to a problem? Please describe.**
A phoneme file is required in order to have any sort of understanding as to how the speech translates to mouth movements.

**Describe the solution you'd like**
Rhubarb must be integrated into our system. 
This will involve the following steps:
- #72 
- #73  

**Describe the feature**
The feature should be able to generate a phoneme file which takes in a text, audio, and language and returns phonemes and timings.

**What aspects are unclear - what needs to be done, or how it should be accomplished?**
We are unsure as to how to integrate a command line tool into our program.

**Additional context**
Issue should be updated as further development occurs.

### LINKED ISSUE #97: Phoneme to MP4 Exploration
**Is this related to a problem? Please describe.**
We don't actually have any method in place to create our video and our animation

**Describe the solution you'd like**
Build a quick program with Typescript that can create an MP4 of our talking lips, or at least have a solid plan in place for how to build that program. 

**Describe the feature**
Preferably having an mp4 of talking mouth would be lovely. 

**What aspects are unclear - what needs to be done, or how it should be accomplished?**
Not sure how to take images and integrate them with rhubarb library to add animation. This will have to be researched.
Here are our two main questions. 
- How do we create an MP4 file with typescript?
- How do we have phoneme mapping to produce animation for our MP4 file?

### LINKED ISSUE #123: FFmpeg integration
**Is this related to a problem? Please describe.**
animations need to be generated
**Describe the solution you'd like**
A clean integration of the FFmpeg command line tool

**Requirements for delivery**
FFmpeg should be able to be ran from a node process

### LINKED ISSUE #124: Mouth Cue to ffmpeg input file
**Is this related to a problem? Please describe.**
Rhubarb generates a list of mouth cues, which need to be used alongside ffmpeg to generate a video. Due to this, the mouth cue list needs to be converted into an input file which is able to be read by ffmpeg

**Describe the solution you'd like**
an asynchronous function which generates an input.txt file in the correct format

**Requirements for delivery**
successful generation of input files

### LINKED ISSUE #140: Convert ffmpeg process to spawn
**Is this related to a problem? Please describe.**
convert to spawn process for increased efficiency

**Describe the solution you'd like**
Similar to rhubarb code

**Requirements for delivery**
ffmpeg needs to run as a spawn process

### LINKED ISSUE #141: Peer testing language selection implementation
**Is this related to a problem? Please describe.**
Create an easy way to select different languages for peer testing

**Describe the solution you'd like**
Change language selections through one simple variable instead of multiple file names

**Requirements for delivery**
simple changes in index 

**Additional context**

### LINKED ISSUE #142: indentation cleanup
**Describe what needs to get done**
Fix inconsistent indentation spacings

**Describe the solution you'd like**
convert spacings from 2 to 4 so that it is consistent project wide & works with tabs - makes it easier to read the code and to write it

### PR DESCRIPTION
**Pull request by LoganParker**
This PR includes the conversion of the ffmpeg code to a spawn process, as well as updates spacing throughout the codebase, maintains consistency of directory changing, and includes code to easily switch between languages without the need to change multiple variables. 
closes #63, closes #97, closes #123, closes #124, closes #140, closes #141, closes #142
^ Some old issues that were closed with previous ffmpeg merges that were not closed. Closing them now.

### CODE REVIEW

**Comment by LoganParker** [comment]
A large number of changes are due to the prettier refactoring that was implemented at the start to keep things properly formatted

File: speaking-portal-project/src/animationFactory/ffmpeg.ts [hunk 1]
```diff
@@ -1,25 +1,38 @@
...
+        `${audio_file}`,
+        '-r',
+        '24',
+        '-pix_fmt',
+        'yuv420p',
+        'output.mp4',
+    ]
+    const ffmpegProc = spawnSync('ffmpeg', args)
+    if (ffmpegProc.stderr) {
+        //throw Error(`${ffmpegProc.stderr}`)
```
Comment (vnajack) [inline speaking-portal-project/src/animationFactory/ffmpeg.ts:27]: When I run with it printing `stderr` to console, this gives a lot of good info. Next semester's exception handling, we can use this to check and make sure the .mp4 is not empty and is good to be returned

**Review by vnajack — APPROVED** [review]
Looks great! Thanks for the code cleanup along with the changes! We'll get more done in exception handling!

File: speaking-portal-project/src/animationFactory/ffmpeg.ts (hunk 1 above)
Comment (LoganParker) [reply speaking-portal-project/src/animationFactory/ffmpeg.ts:27]: Yes! I actually forgot to uncomment this, so feel free to suggest that change and make it!

Comment (LoganParker) [reply speaking-portal-project/src/animationFactory/ffmpeg.ts:27]: We can also print it out when we have a testing mode enabled or something along those lines

Comment (edouarde1) [reply speaking-portal-project/src/animationFactory/ffmpeg.ts:27]: yes! There might be a a workaround this issue. I think we are capturing the output and not errors. We can looking into this. The same issue exists for the phoneme factory too!

File: speaking-portal-project/src/index.ts [hunk 2]
```diff
@@ -1,37 +1,49 @@
...
-    try{
+    let audio_file_name: string = audio_file_options[language_selection]
+    let text_file_name: string = text_file_options[language_selection]
+    let selected_language: string = recognizer_options[language_selection]
+
+    if (!audio_file_name || !text_file_name || !selected_language) {
+        console.log('Language Selection out of bounds\nLanguage set to English (U.S.)')
+        audio_file_name = audio_file_options[0]
+        text_file_name = text_file_options[0]
+        selected_language = recognizer_options[0]
```
Comment (edouarde1) [inline speaking-portal-project/src/index.ts:28]: Good set-up, makes it really easy for the user to test different languages

File: speaking-portal-project/src/phonemeFactory/mouthCueProcessor.ts [hunk 3]
```diff
@@ -1,39 +1,38 @@
...
+        })
+        /* Lines 17-20 for testing purposes only
+        const contents = await fs.promises.readFile(outputPath, 'utf-8')
+        console.log(contents);
+    
+        return contents
+        */
+    } catch (err) {
+        console.log(err)
+        return 'Something went wrong'
```
Comment (edouarde1) [inline speaking-portal-project/src/phonemeFactory/mouthCueProcessor.ts:23]: haha will make our error handling a bit more concise (soon)

File: speaking-portal-project/src/phonemeFactory/mouthCueProcessor.ts [hunk 4]
```diff
@@ -1,39 +1,38 @@
...
-    console.log(contents);
-
-    return contents
-    */
-  } catch (err) {
-    console.log(err)
-    return 'Something went wrong'
-  }
+    try {
+        await fs.promises.writeFile(outputPath, generateMouthData(mouthCues), {
```
Comment (edouarde1) [inline speaking-portal-project/src/phonemeFactory/mouthCueProcessor.ts:12]: From my understanding we are writing the output from generateMouthData(mouthCues) into the 'outputPath' to a file ?

**Review by edouarde1 — APPROVED** [review]

File: speaking-portal-project/src/phonemeFactory/mouthCueProcessor.ts (hunk 4 above)
Comment (LoganParker) [reply speaking-portal-project/src/phonemeFactory/mouthCueProcessor.ts:12]: Yes! Gets written to input.txt which ffmpeg reads in order to know which frames to place at what times

**Comment by mkuelker** [comment]
Closing this branch

**Pull request merged**